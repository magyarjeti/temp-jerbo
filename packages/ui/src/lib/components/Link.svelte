<script lang="ts">
	import isAbsoluteUrl from '../utils/isAbsoluteUrl';
	import isLocalUrl from '../utils/isLocalUrl';
	import type { HTMLAnchorAttributes } from 'svelte/elements';
	import type { Snippet } from 'svelte';
	import { concatClasses } from '@packages/toolbox/concatClasses';

	type Props = {
		disabled?: boolean | null | undefined;
		allowReferer?: boolean;
		children?: Snippet;
	} & HTMLAnchorAttributes;

	let { disabled, allowReferer = false, children, ...attrs }: Props = $props();

	let isExternalLink = $derived(isAbsoluteUrl(attrs.href) && !isLocalUrl(attrs.href));
	const classes = $derived(concatClasses(attrs.class, disabled && 'disabled opacity-38'));
</script>

{#if attrs.href}
	<a
		target={isExternalLink ? '_blank' : undefined}
		rel={isExternalLink ? ('noopener' + !allowReferer ? ' noreferrer' : '') : undefined}
		aria-disabled={disabled}
		{...attrs}
		class={classes}>{@render children?.()}</a
	>
{:else}
	<span class={classes} aria-disabled={disabled}>{@render children?.()}</span>
{/if}
