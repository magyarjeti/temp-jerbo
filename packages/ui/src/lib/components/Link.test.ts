import { render, screen } from '@testing-library/svelte';
import { expect, test } from 'vitest';

import Link from './Link.svelte';

test('link snapshot', () => {
	const { container } = render(Link, { props: { href: 'http://example.com' } });
	expect(container).toMatchSnapshot();
});

test('link is external', () => {
	render(Link, { props: { href: 'http://example.com' } });
	const anchor = screen.getByRole('link');
	expect(anchor.getAttribute('href')).toBe('http://example.com');
	expect(anchor.getAttribute('target')).toBe('_blank');
	expect(anchor.getAttribute('rel')).toBe('noopener noreferrer');
});

test('link is internal', () => {
	render(Link, {
		props: { href: '/test' },
	});
	const anchor = screen.getByRole('link');
	expect(anchor.getAttribute('href')).toBe('/test');
});
