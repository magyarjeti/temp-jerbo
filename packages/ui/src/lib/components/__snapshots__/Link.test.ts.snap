// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`link snapshot 1`] = `
<div>
  <a
    href="http://example.com"
    rel="noopener noreferrer"
    target="_blank"
  />
  
  
</div>
`;
